File: odemapster-r2rml-rdb/src/main/java/es/upm/fi/dia/oeg/obdi/wrapper/r2rml/rdb/model/R2RMLPredicateMap.java
package es.upm.fi.dia.oeg.obdi.wrapper.r2rml.rdb.model;

import com.hp.hpl.jena.rdf.model.Resource;

import es.upm.fi.dia.oeg.morph.base.Constants;
import es.upm.fi.dia.oeg.obdi.wrapper.r2rml.rdb.exception.R2RMLInvalidTermMapException;
import es.upm.fi.dia.oeg.obdi.wrapper.r2rml.rdb.model.R2RMLTermMap.TermMapType;

public class R2RMLPredicateMap extends R2RMLTermMap {
	
//	public R2RMLPredicateMap(String constantValue) {
//		super(TermMapPosition.PREDICATE, constantValue);
//		super.setTermType(Constants.R2RML_LITERAL_URI());
//	}
	
	public static R2RMLPredicateMap create(Resource resource, R2RMLTriplesMap owner) 
			throws R2RMLInvalidTermMapException {
		R2RMLPredicateMap pm = new R2RMLPredicateMap();
		pm.parse(resource, owner);
		return pm;
	}

	public static R2RMLPredicateMap create(String constantValue) {
		R2RMLPredicateMap pm = new R2RMLPredicateMap();
		pm.termMapType = TermMapType.CONSTANT;
		pm.constantValue = constantValue;
		pm.termType = Constants.R2RML_IRI_URI();
		return pm;
	}	
}

File: morph-r2rml/src/main/scala/es/upm/dia/fi/oeg/morph/r2rml/model/R2RMLPredicateMap.scala
package es.upm.dia.fi.oeg.morph.r2rml.model

import scala.collection.JavaConversions._

import es.upm.fi.dia.oeg.morph.base.Constants
import es.upm.fi.dia.oeg.obdi.core.model.AbstractConceptMapping
import com.hp.hpl.jena.rdf.model.Resource

class R2RMLPredicateMap() extends R2RMLTermMap {
	
	def this(resource:Resource) = {
		this();
		this.parse(resource);
		this.resource = resource;
	}

	def this(constantValue:String) = {
		this();
		this.termMapType = Constants.MorphTermMapType.ConstantTermMap;
		this.constantValue = constantValue;
		this.termType = Constants.R2RML_IRI_URI;
	}	  
}

object R2RMLPredicateMap {
	def extractPredicateMaps(resource:Resource) 
	: List[R2RMLPredicateMap] = {
		val pmStatements = resource.listProperties(Constants.R2RML_PREDICATEMAP_PROPERTY);
		val predicateMaps1 = if(pmStatements != null) {
			pmStatements.toList().flatMap(pmStatement => {
				if(pmStatement != null) {
					val predicateMapResource = pmStatement.getObject().asInstanceOf[Resource];
					val pm = new R2RMLPredicateMap(predicateMapResource);
					Some(pm);
				} else {
				  None
				}			  
			});
		} else {
		  Nil
		}
		
		val pStatements = resource.listProperties(Constants.R2RML_PREDICATE_PROPERTY);
		val predicateMaps2 = if(pStatements != null) {
			pStatements.toList().flatMap(predicateStatement => {
				if(predicateStatement != null) {
					val constantValueObject = predicateStatement.getObject().toString();
					val predicateMap = new R2RMLPredicateMap(constantValueObject);
					Some(predicateMap);
				} else {
				  None
				}			  
			}); 
		} else {
		  Nil
		}

		val predicateMaps = predicateMaps1.toList ::: predicateMaps2.toList;
		predicateMaps
	}  
}